[CEP 26] Identifying Packages and Channels in the conda Ecosystem (#116)
conda/ceps: cep-XXXX.md

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]/<subdir>/repodata.json
```

with `<scheme>`, `<authority>` and `<path>` defined by [RFC
3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2).

Taken the channel definition above, the base URL without trailing slashes is thus:

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]
```

For example, given `https://conda.anaconda.org/conda-forge/noarch/repodata.json`, the part leading
to `noarch/repodata.json` and thus base URL is `https://conda.anaconda.org/conda-forge/conda-forge`. For local repodata such as `file:///home/username/channel/noarch/repodata.json`, the
channel base URL is `file:///home/username/channel`.

For convenience, the channel _name_ is defined as the concatenation of `scheme`, `authority` and
`path` components. `scheme` MAY be omitted for `http`, `https` and `file`. At least one of
`authority` or `path` SHOULD be present. In their absence, the channel name MUST be considered
empty, regardless the scheme. Empty channel names SHOULD NOT be used.

When present, each path component SHOULD only contain lowercase ASCII letters, numbers, underscores

> and dashes. They SHOULD start and end with a letter or a number.

Reviewer: Is there a reason we do not allow consecutive separators at all in conda package names, but we do allow them in channel path components?

Reviewer: if we want to cater to OCI, then this will need to be restricted to the OCI name regex.

Reviewer: The regex needed by OCI for labels and channel path components is

```
^[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*$
```

which is the `v1.*` OCI regex for container `<name>`s. 

I do not have stats on how many labels violate this on anaconda.org. The number of channels from usernames/orgs that violate it is around 0.4% which is small enough that I excluded them in the OCI CEP. 

Maybe we simply exclude them here too?

Reviewer: Also this is the stat from @chenghlee 

> As of 2025-03-12T19:00Z, of the ~1.9M channel/org/names on anaconda.org:
> 
> 7,219 violate the regex ^[a-z0-9]+((-|_|.)[a-z0-9]+)*$;
> 98 violate the regex ^[a-z0-9][a-z0-9_.-]*$ (allowing channel names to end with _, ., or -); and
> 6 violate ^[a-z0-9_][a-z0-9_.-]*$ (allowing channel names to start with _)
> Of those six, five start with ., and the other starts with ~. Which I would claim, we want to disallow in any case (since they correspond to hidden files, directory traversal, or a user home directory in *NIX).
> 
> Edit: With the caveat that I used case-insensitive matching when evaluating the above. But I'm okay with requiring conda-compatible software to normalize channel names by lower-casing.

So the requirement that things start and end with a letter already excludes the 0.4% regardless of the issue of how many separators appear in a row. Thus we probably should just use the OCI name regex and be done with it?

Author: Ok, I can add this as a SHOULD for now. Or do you want it as MUST?

Reviewer: Looking at this more, I think we should go all one way (very liberal) or the other (restrict to OCI). In between the those cases doesn't make much sense to me. 

So I think we specify that channel path components MUST match `^[a-z0-9_][[a-z0-9_.-]$` and let the OCI spec exclude more.

Author: Works for me.

The maximum length of a subdir name MUST NOT exceed 32 characters.

#### Label names

Channel label names MUST only consist of ASCII letters, digits, underscores, hyphens, forward slashes, periods, colons and whitespace. They MUST start with a letter. They MUST match this regex: `^[a-zA-Z][0-9a-zA-Z_\-\.\/:\s]*`. Even if allowed, label names SHOULD NOT contain any whitespace.

The label `nolabel` is reserved and MUST only be used for conda packages which have no other labels. In other words, in the space of labels, the empty set is represented by the labels `nolabel`.

The maximum length of a label name SHOULD NOT exceed 128 characters.

## Backwards compatibility

The conda channel, subdir, label, and package name regexes are backwards compatible with the current `conda` implementation (25.3) and all existing packages on the `defaults` and `conda-forge` channels, except the `__anaconda_core_depends` package on the `defaults` channel.

> The regex for labels above was pulled from an anaconda.org error message describing the set of valid labels. Finally, anaconda.org usernames/channel names are case insensitive and cannot begin with an underscore.

Reviewer: It would be good to collect the stats from @chenghlee on anaconda.org and put them as text here.

Virtual package names MUST follow this other regex: `^__[a-z0-9][._-]?([a-z0-9]+(\.|-|_|$))*$`.

In all cases, the maximum length of a package name MUST NOT exceed 128 characters.

#### Version strings

Version strings MUST only consist of digits, periods, lowercase ASCII letters, underscores, plus
symbols, and exclamation marks. Additional rules apply but are out of scope in this CEP and will be
discussed separately.

The maximum length of a version string MUST NOT exceed 128 characters.

#### Build strings

Builds strings MUST only consist of ASCII letters, numbers, periods, plus symbols, and underscores. They MUST match this regex `^[a-zA-Z0-9_\.+]+$`.

The maximum length of a build string MUST NOT exceed 128 characters.

#### Filenames

Distributable conda artifacts MUST have a filename following this scheme:

```text
<package name>-<version string>-<build string>.<extension>
```

Virtual conda packages do not exist on disk and SHOULD NOT need filename standardization.

#### Distribution strings

A "distribution string" MAY be used to identify a package artifact without specifying the extension or the channel. It MUST match the following syntax:

```text

> <package name>-<version string>-<build string>

Reviewer: We need to use space-separated and/or `=`-separated forms of this as valid matchspecs in recipes. We may want to be clearer about the scope of this section or add a carve out for matchspecs

Author: Yea this is not a MatchSpec, is a dist str, which is not allowed anywhere, just a convention like filenames are. I'll clarify.

#### Channel base URLs and names

The base URL for the arbitrary location of a repodata file is defined as:

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]/<subdir>/repodata.json
```

with `<scheme>`, `<authority>` and `<path>` defined by [RFC
3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2).

Taken the channel definition above, the base URL without trailing slashes is thus:

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]
```

For example, given `https://conda.anaconda.org/conda-forge/noarch/repodata.json`, the part leading
to `noarch/repodata.json` and thus base URL is `https://conda.anaconda.org/conda-forge/conda-forge`. For local repodata such as `file:///home/username/channel/noarch/repodata.json`, the
channel base URL is `file:///home/username/channel`.

For convenience, the channel _name_ is defined as the concatenation of `scheme`, `authority` and
`path` components. `scheme` MAY be omitted for `http`, `https` and `file`. At least one of
`authority` or `path` SHOULD be present. In their absence, the channel name MUST be considered

> empty, regardless the scheme. Empty channel names SHOULD NOT be used.

Reviewer: I may have made this mistake in the OCI CEP too. What you mean when you say the scheme MAY be omitted is that we allow things like `conda-forge`. Literally removing the scheme above would look like `://conda-forge` which is not what we mean.

Author: Oops, good point. I'll add a note.

The label `nolabel` is reserved and MUST only be used for conda packages which have no other labels. In other words, in the space of labels, the empty set is represented by the labels `nolabel`.

The maximum length of a label name SHOULD NOT exceed 128 characters.

## Backwards compatibility

The conda subdir and package name regexes are backwards compatible with the current `conda` implementation (25.3) and all existing packages on the `defaults` and `conda-forge` channels, except for the `__anaconda_core_depends` package on the `defaults` channel.

The regex for labels was pulled from an anaconda.org error message describing the set of valid labels.

As of 2025-03-12T19:00Z, of the ~1.9M channel names on anaconda.org:

- 7,219 violate the regex `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`;
- 98 violate the regex `^[a-z0-9][a-z0-9_.-]*$` (allowing channel names to end with `_`, `.`, or `-`); and
- 6 violate `^[a-z0-9_][a-z0-9_.-]*$` (allowing channel names to start with `_`). Of those six, five start with `.`, and the other starts with `~`.

> Given the low percentage (~0.4%) of mismatches, authors have decided to ignore those entries and err on the side of strictness to maximize compatibility with other storage solutions that have more restrictive naming requirements.

Reviewer: May not be true given our discussion above.

The label `nolabel` is reserved and MUST only be used for conda packages which have no other labels. In other words, in the space of labels, the empty set is represented by the labels `nolabel`.

The maximum length of a label name SHOULD NOT exceed 128 characters.

## Backwards compatibility

The conda subdir and package name regexes are backwards compatible with the current `conda` implementation (25.3) and all existing packages on the `defaults` and `conda-forge` channels, except for the `__anaconda_core_depends` package on the `defaults` channel.

The regex for labels was pulled from an anaconda.org error message describing the set of valid labels.

As of 2025-03-12T19:00Z, of the ~1.9M channel names on anaconda.org:

- 7,219 violate the regex `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`;
- 98 violate the regex `^[a-z0-9][a-z0-9_.-]*$` (allowing channel names to end with `_`, `.`, or `-`); and
- 6 violate `^[a-z0-9_][a-z0-9_.-]*$` (allowing channel names to start with `_`). Of those six, five start with `.`, and the other starts with `~`.

> Given the low percentage (~0.4%) of mismatches, authors have decided to ignore those entries and err on the side of strictness to maximize compatibility with other storage solutions that have more restrictive naming requirements.

Reviewer: ```suggestion
The authors have excluded the channel names in the last case that start with `.` or `~` given possible security implications. A low percentage, ~0.4%, of channels do not match the recommendations for channel names above, but are allowed.
```

# CEP XXXX - Names in conda packages and channels

<table>
<tr><td> Title </td><td> CEP XXXX - Names in conda packages and channels </td>
<tr><td> Status </td><td> Draft </td></tr>

> <tr><td> Author(s) </td><td> Jaime Rodríguez-Guerra &lt;jaime.rogue@gmail.com&gt; <br /> Matthew R. Becker &lt;becker.mr@gmail.com&gt;</td></tr>

Author: @chenghlee, would you like to be part of the author list? You've helped a lot!

```text
<package name>-<version string>-<build string>
```

Distribution strings apply to both distributable and virtual packages. They are used as the name of the
directories where artifacts are extracted in the package cache, for example.

> Note: Despite the similarity, distribution strings are not `MatchSpec`-like specifiers and
> MUST NOT be used as such.

### Identifying channels

A conda channel is defined as a URL where one can find one or more `repodata.json` files arranged in one
subdirectory (_subdir_) each. `noarch/repodata.json` MUST be present to consider the parent location a channel.

#### Channel base URLs and names

The base URL for the arbitrary location of a repodata file is defined as:

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]/<subdir>/repodata.json
```

with `<scheme>`, `<authority>` and `<path>` defined by [RFC
3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2).

Taken the channel definition above, the base URL without trailing slashes is thus:

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]
```

For example, given `https://conda.anaconda.org/conda-forge/noarch/repodata.json`, the part leading

> to `noarch/repodata.json` and thus base URL is `https://conda.anaconda.org/conda-forge/conda-forge`. For

Author: Oops, yes, this is a typo.

local repodata such as `file:///home/username/channel/noarch/repodata.json`, the
channel base URL is `file:///home/username/channel`.

For convenience, the channel _name_ is defined as the concatenation of `scheme`, `authority` and
`path` components. The `<scheme>://` part MAY be omitted if `scheme` is one of `http`, `https` or

> `file`. At least one of `authority` or `path` SHOULD be present. In their absence, the channel name

Reviewer 2: > The `<scheme>://` part MAY be omitted if `scheme` is one of `http`, `https` or
`file`.

Should we restrict this to just `file`? It seems odd to me that if I do something like `mkdir -pv conda.anaconda.org/mychannel/noarch`, multiple conda-compatible clients now have to figure out and agree if the schema-less argument `--channel conda.anaconda.org/mychannel/` refers to a local relative (`file://`-like) path or a HTTP(S) service.

Author: Added a new section for this only. Let me know what you think.

Reviewer: Looks good. @chenghlee @jaimergp Do we know what rules `conda` uses when doing this type of inference? Do they match what is in the CEP? The rules don't have to match, but it'd be good to compare so we know what is currently done.

Author: Not really, I think.

For environment names (not channels), there are two CLI parts:

- `conda create` disambiguates via `-n` (name) and `-p` (path, prefix).
- `conda activate` does not, and instead relies on [this logic](https://github.com/conda/conda/blob/866f84041cb40a4122fe5e8154d67eb28817884a/conda/activate.py#L344-L353). This is similar to what we propose here.

For channels, it's a bit more [involved](https://github.com/conda/conda/blob/866f84041cb40a4122fe5e8154d67eb28817884a/conda/models/channel.py#L442-L472). They strip path components until there's a match in `custom_channels`, or give up and append it to `channel_alias` (conda.anaconda.org).

Reviewer: ok LGTM.

MUST be considered empty, regardless the scheme. Empty channel names SHOULD NOT be used.

When present, each path component MUST only contain lowercase ASCII letters, numbers, underscores,
periods, and dashes. They MUST NOT start with a period or a dash. They SHOULD start and end with a
letter or a number. If present, each path component MUST match this regex:

```re
^[a-z0-9_][a-z0-9_.-]*$

> ```

Reviewer 2: This might create a potential problem with `file://` URLs, especially when it comes to user home directories.

Author: I'll add an exception for `file://` channels, downgrading it to a SHOULD.

The maximum length of an individual path component in a channel base URL MUST NOT exceed 128 characters.
The maximum length of a channel base URL SHOULD NOT exceed 256 characters.

#### Subdir names

Channel subdir names MUST either be the literal `noarch` or a string following the syntax `{os}-{arch}`,
where `{os}` and `{arch}` MUST only consist of lowercase ASCII letters and numbers. Non-`noarch` subdirs
MUST match this regex: `^[a-z0-9]+-[a-z0-9]+$`.

The maximum length of a subdir name MUST NOT exceed 32 characters.

#### Label names

Channel label names MUST only consist of ASCII letters, digits, underscores, hyphens, forward slashes,

> periods, colons and whitespace. They MUST start with a letter. They MUST match this regex:

Reviewer 2: I'm inclined to disallow colons, since they're disallowed on CIFS and Windows (except for drive letters) and get interpreted as directory delimiters on macOS.

Author: Applied suggestions.

`^[a-zA-Z][0-9a-zA-Z_\-\.\/:\s]*`. Even if allowed, label names SHOULD NOT contain any whitespace.

The label `nolabel` is reserved and MUST only be used for conda packages which have no other labels.
In other words, in the space of labels, the empty set is represented by the labels `nolabel`.

> 

Reviewer: ```suggestion

A URL for a package, repodata, etc. without a label component MUST be assumed to have the default label `main`.

```

As of 2025-03-12T19:00Z, of the ~1.9M channel names on anaconda.org:

- 7,219 violate the regex `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`;
- 98 violate the regex `^[a-z0-9][a-z0-9_.-]*$` (allowing channel names to end with `_`, `.`, or `-`); and
- 6 violate `^[a-z0-9_][a-z0-9_.-]*$` (allowing channel names to start with `_`).
  Of those six, five start with `.`, and the other starts with `~`.

See [this comment](https://github.com/conda/ceps/pull/116#discussion_r1992154574) for more details.
The authors have excluded the channel names in the last case that start with `.` or `~` given possible
security implications. A low percentage, ~0.4%, of channels do not match the recommendations for channel
names above, but are allowed.

The maximum lengths allowed for the different fields have been chosen so the resulting path components
(directory names, filenames) comfortably fit in a the 255-char maximum limit some filesystems impose.
As of 2025-03-01T13:00Z, there are no violations of these limits in any of the packages published for
`conda-forge`, `bioconda` and `defaults`. See
[this comment](https://github.com/conda/ceps/pull/116#issuecomment-2763392999) and

> [this comment](https://github.com/conda/ceps/pull/116#issuecomment-2759130187)for more details.

Reviewer: ```suggestion
[this comment](https://github.com/conda/ceps/pull/116#issuecomment-2759130187) for more details.
```

- Otherwise, it SHOULD be understood as a `http[s]://` URL. The tool SHOULD assume a default scheme and authority (e.g. `https://conda.anaconda.org`), and take the rest as a path component.

#### Subdir names

Channel subdir names MUST either be the literal `noarch` or a string following the syntax `{os}-{arch}`,
where `{os}` and `{arch}` MUST only consist of lowercase ASCII letters and numbers. Non-`noarch` subdirs
MUST match this regex: `^[a-z0-9]+-[a-z0-9]+$`.

The maximum length of a subdir name MUST NOT exceed 32 characters.

#### Label names

Channel label names MUST only consist of ASCII letters, digits, underscores, hyphens, forward slashes,
periods, and whitespace. They MUST start with a letter. They MUST match this regex:
`^[a-zA-Z][0-9a-zA-Z_\-\.\/\s]*`. Even if allowed, label names SHOULD NOT contain any whitespace.

The label `nolabel` is reserved and MUST only be used for conda packages which have no other labels.
In other words, in the space of labels, the empty set is represented by the labels `nolabel`.

A URL for a package, repodata, etc. without a label component MUST be assumed to have the default label `main`.

> The maximum length of a label name MUST NOT exceed 128 characters.

Reviewer: ```suggestion

The maximum length of a label name MUST NOT exceed 128 characters.
```

# CEP XXXX - Names in conda packages and channels

<table>

> <tr><td> Title </td><td> CEP XXXX - Names in conda packages and channels </td>

Reviewer: ```suggestion
<tr><td> Title </td><td> CEP XXXX - Conda Channel URLs </td>
```

<tr><td> Status </td><td> Draft </td></tr>
<tr><td> Author(s) </td><td>
  Jaime Rodríguez-Guerra &lt;jaime.rogue@gmail.com&gt; <br />
  Matthew R. Becker &lt;becker.mr@gmail.com&gt; <br />
  Cheng H. Lee &lt;clee@anaconda.com&gt;
</td></tr>
<tr><td> Created </td><td> Mar 11, 2025</td></tr>
<tr><td> Updated </td><td> Apr 1, 2025</td></tr>
<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/116 </td></tr>
<tr><td> Implementation </td><td> N/A </td></tr>
</table>

## Abstract

> This CEP aims to standardize names and other strings used to identify packages, artifacts and
> channels in the conda ecosystem.

Reviewer: ```suggestion
This CEP aims to standardize conda channel URLs, package names, and other strings used to identify packages, artifacts and channels in the conda ecosystem.
```

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
"RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as
described in [RFC2119][RFC2119] when, and only when, they appear in all capitals, as shown here.

More specifically, violations of a MUST or MUST NOT rule MUST result in an error. Violations of the
rules specified by any of the other all-capital terms MAY result in a warning, at discretion of the
implementation.

### Identifying package artifacts

The conda ecosystem distinguishes between two types of packages:

- Distributable package names: represented by a concrete, downloadable, extractable conda artifact.

> - Virtual package names: not backed by any physical artifact. They only exist on the client side.

Reviewer 3: ```suggestion
- Virtual package names: not backed by any concrete artifact. They only exist on the client side.
```

Align for consistency across definitions

#### Package names

A distributable package name MUST only consist of lowercase ASCII letters, numbers, hyphens, periods
and underscores. It MUST start with a letter, a number, or a single underscore. It MUST NOT include
two consecutive separators (hyphen, period, underscore).

Virtual package names MUST only consist of lowercase ASCII letters, numbers, hyphens, periods and
underscores. They MUST NOT use two consecutive separators, with one exception: they MUST start with
two underscores.

> This means that distributable package names MUST match the following case-insensitive
> regex: `^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`.

Reviewer 3: Nit: let's move this before `Virtual packages` (line no. 45) for readability too

```suggestion
Distributable package names MUST match the following case-insensitive
regex: `^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`.
```

#### Version strings

Version strings MUST only consist of digits, periods, lowercase ASCII letters, underscores, plus
symbols, and exclamation marks. Additional rules apply but are out of scope in this CEP and will be
discussed separately.

The maximum length of a version string MUST NOT exceed 64 characters.

#### Build strings

Builds strings MUST only consist of ASCII letters, numbers, periods, plus symbols, and underscores.
They MUST match this regex `^[a-zA-Z0-9_\.+]+$`.

The maximum length of a build string MUST NOT exceed 64 characters.

#### Artifact extensions

Artifact extensions MUST only consist of lowercase ASCII letters, numbers and periods. They must start
and end with a letter or a number. They MUST NOT include two consecutive periods. They MUST match this
regex `^[a-z0-9](\.?[a-z0-9])*$`.

The maximum length of a file extension MUST NOT exceed 16 characters.

> The conda ecosystem currently recognizes two artifact extensions: `tar.bz2` and `conda`, versioned
`v1` and `v2` respectively.

#### Filenames

Distributable conda artifacts MUST have a filename following this scheme:

```text
<package name>-<version string>-<build string>.<extension>
```

> The maximum length of a filename MUST NOT exceed 192 characters.

Reviewer 3: ```suggestion
The maximum length of a filename MUST NOT exceed 211 characters.
```

Virtual conda packages do not exist on disk and SHOULD NOT need filename standardization.

#### Distribution strings

A "distribution string" MAY be used to identify a package artifact without specifying the extension
or the channel. It MUST match the following syntax:

```text
<package name>-<version string>-<build string>
```

Distribution strings apply to both distributable and virtual packages. They are used as the name of the
directories where artifacts are extracted in the package cache, for example.

> Note: Despite the similarity, distribution strings are not `MatchSpec`-like specifiers and

> > MUST NOT be used as such.

Reviewer 3: This would be best moved _before_ filenames as the latter inherit from the `Distribution strings` spec

their absence, the channel name MUST be considered empty, regardless the scheme. Empty channel
names SHOULD NOT be used.

When the scheme and authority fields are missing, the full URL can be inferred with these rules:

- If the channel name matches the regex `^\.{0,2}[/\\].*$`, or if it matches the regex `^[A-Z]:([\\/].*)?$` (for Windows drives), it SHOULD be understood as the path component of a `file://` URL.
- Otherwise, it SHOULD be understood as a `http[s]://` URL. The tool SHOULD assume a default scheme and authority (e.g. `https://conda.anaconda.org`), and take the rest as a path component.

#### Subdir names

Channel subdir names MUST either be the literal `noarch` or a string following the syntax `{os}-{arch}`,
where `{os}` and `{arch}` MUST only consist of lowercase ASCII letters and numbers. Non-`noarch` subdirs
MUST match this regex: `^[a-z0-9]+-[a-z0-9]+$`.

The maximum length of a subdir name MUST NOT exceed 32 characters.

#### Label names

Channel label names MUST only consist of ASCII letters, digits, underscores, hyphens, forward slashes,
periods, and whitespace. They MUST start with a letter. They MUST match this regex:

> `^[a-zA-Z][0-9a-zA-Z_\-\.\/\s]*`. Even if allowed, label names SHOULD NOT contain any whitespace.

Reviewer: @chenghlee @jaimergp as long as we're deviating from anaconda.org here (with the removal of `:`s), we might as well remove whitespace `\s` as well?

Author: I'm ok with that.

Author: I can also add back `:` and include it in the SHOULD NOT part.

Reviewer: I think both `:` and `\s` are bad ideas so I am happy removing them both.

Author: Farewell to \s then!

> # CEP XXXX - Names in conda packages and channels

Reviewer: ```suggestion
# CEP XXXX - Identifying Packages and Channels in the conda Ecosystem
```

letter or a number. If present, each path component MUST match this regex:

```re
^[a-z0-9_][a-z0-9_.-]*$
```

For `file://`-based channel URLs, the path component rules MAY be understood as recommendations
only.

The maximum length of an individual path component in a channel base URL MUST NOT exceed 128
characters. The maximum length of a channel base URL SHOULD NOT exceed 256 characters.

#### Channel names

For convenience, the channel _name_ is defined as the concatenation of `scheme`, `authority` and
`path` components of a channel URL. At least one of `authority` or `path` SHOULD be present. In
their absence, the channel name MUST be considered empty, regardless the scheme. Empty channel
names SHOULD NOT be used.

When the scheme and authority fields are missing, the full URL can be inferred with these rules:

- If the channel name matches the regex `^\.{0,2}[/\\].*$`, or if it matches the regex
  `^[A-Z]:([\\/].*)?$` (for Windows drives), it SHOULD be understood as the path component of a
  `file://` URL.
- Otherwise, it SHOULD be understood as a `http[s]://` URL. The tool SHOULD assume a default scheme

>   and authority (e.g. `https://conda.anaconda.org`), and take the rest as a path component.

Reviewer 2: How about
> Otherwise, the tool SHOULD provide a user-configurable mechanism to prepend a default scheme and authority, with provided channel name taken as the rest of the path component. At the time of this CEP's writing, most tool assume the default URL scheme and authority to be `https://conda.anaconda.org`.

I think this better captures the intent (i.e., support the `-c`/`--channel` option as they exist today), while recognizing that condarc already provides mechanisms to have the default schema not be `http[s]://` for certain channels.

Version strings MUST only consist of digits, periods, lowercase ASCII letters, underscores, plus
symbols, and exclamation marks. Additional rules apply but are out of scope in this CEP and will be
discussed separately.

The maximum length of a version string MUST NOT exceed 64 characters.

#### Build strings

Builds strings MUST only consist of ASCII letters, numbers, periods, plus symbols, and underscores.
They MUST match this regex `^[a-zA-Z0-9_\.+]+$`.

The maximum length of a build string MUST NOT exceed 64 characters.

#### Artifact extensions

Artifact extensions MUST only consist of lowercase ASCII letters, numbers and periods. They must
start and end with a letter or a number. They MUST NOT include two consecutive periods. They MUST
match this regex `^[a-z0-9](\.?[a-z0-9])*$`.

The maximum length of a file extension MUST NOT exceed 16 characters.

> The conda ecosystem currently recognizes two artifact extensions: `tar.bz2` and `conda`,
versioned `v1` and `v2` respectively.

#### Distribution strings

A "distribution string" MAY be used to identify a package artifact without specifying the extension
or the channel. It MUST match the following syntax:

```text

> <package name>-<version string>-<build string>

Reviewer: We may want to add a possible `<subdir>` part here too?

```suggestion
[<subdir>/]<package name>-<version string>-<build string>
```

cc @jaimergp

Author: Good point. Let me rephrase the filename bit too so it doesn't add confusion, though.

with `<scheme>`, `<authority>` and `<path>` defined by [RFC
3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2).

Taken the channel definition above, the base URL without trailing slashes is thus:

```text
<scheme>://[<authority>][/<path>/][/label/<label name>]
```

For example, given `https://conda.anaconda.org/conda-forge/noarch/repodata.json`, the part leading
to `noarch/repodata.json` and thus base URL is `https://conda.anaconda.org/conda-forge`. For local
repodata such as `file:///home/username/channel/noarch/repodata.json`, the channel base URL is
`file:///home/username/channel`.

When present, each path component MUST only contain lowercase ASCII letters, numbers, underscores,
periods, and dashes. They MUST NOT start with a period or a dash. They SHOULD start and end with a
letter or a number. If present, each path component MUST match this regex:

```re
^[a-z0-9_][a-z0-9_.-]*$
```

For `file://`-based channel URLs, the path component rules MAY be understood as recommendations
only.

The maximum length of an individual path component in a channel base URL MUST NOT exceed 128
characters. The maximum length of a channel base URL SHOULD NOT exceed 256 characters.

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Reviewer: This is not formally a requirement that can always be met, since the `subdirs` change. For example a channel might be `blah/bar/foo-128` and be valid. Then if someone comes along and declares a subdir `foo-128` later for a new platform and arch, we now have channels that were valid, but are now invalid. This seems like a rare edge case, but I wonder if we should think a bit more on this?

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Author: We can downgrade to SHOULD and add some language so folks avoid those pitfalls. e.g. don't register a channel named `linux-128` just in case 😂

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Reviewer: What is your proposal on changing matchspecs? We'd need a separator besides `/` or to force more items into the key=value bits.

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Author: We would deprecate the `channel/subdir` notation and instead point users to use `subdir=<value>` in the bracket specs (which is supported already):

```python
>>> from conda.models.match_spec import MatchSpec
>>> MatchSpec("conda-forge/linux-64::python")
MatchSpec("conda-forge/linux-64::python")
>>> MatchSpec("conda-forge::python[subdir=linux-64]")
MatchSpec("conda-forge/linux-64::python")
>>> MatchSpec("conda-forge/linux-64::python") == MatchSpec("conda-forge::python[subdir=linux-64]")
True
```

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Reviewer: Right. So to be clear these two things can be parsed reliably

- `<channel base URL>(/label/<label>)/<subdir>/<package or repodata or other file>`
- `<channel base URL>(/label/<label>)`

but this these are ambiguous

- `<channel base URL>(/label/<label>)/<subdir>`
- `conda-forge/label/linux-64`

The only thing we actually need is a way to tell the parser that even if a path component matches a subdir, to ignore that match.

Could we allow users to optionally use percent-encoding to encode the individual parts to resolve any ambiguity?

For example, suppose a channel is made that is

- `foo/bar/linux-128`

Then later, the subdir `linux-128` is added. We could let users denote this channel as

- `foo/bar/linux%2D128/label/osx%2D64/osx-64`
- `foo/bar/linux%2D128/osx-64`

with the idea that anything that is percent encoded is meant to be decoded when used as a URL, but to remain encoded when parsing.

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Author: In case of ambiguity we can also check for the existence of `my-channels/linux-64/repodata.json` and decide based on that, but doing IO to resolve parsing ambiguity sounds so wrong.

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Author: Yep, I thought about that but then it wouldn't actually correspond to the path on disk 🥲

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Author: Maybe we just recommend against it and defer to tools how to resolve the ambiguity. Future CEPs could resolve this by proposing a different file tree for channels.

> To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL MUST NOT match any `subdir` identifiers.

Reviewer: ```suggestion
To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL SHOULD NOT match any `subdir` identifiers. If it does, the behavior in this ambiguous
case is not defined and implementation dependent.
```

# CEP XXXX - Identifying Packages and Channels in the conda Ecosystem

<table>
<tr><td> Title </td><td> CEP XXXX - Identifying Packages and Channels in the conda Ecosystem </td>
<tr><td> Status </td><td> Draft </td></tr>
<tr><td> Author(s) </td><td>
  Jaime Rodríguez-Guerra &lt;jaime.rogue@gmail.com&gt; <br />
  Matthew R. Becker &lt;becker.mr@gmail.com&gt; <br />
  Cheng H. Lee &lt;clee@anaconda.com&gt;
</td></tr>
<tr><td> Created </td><td> Mar 11, 2025</td></tr>

> <tr><td> Updated </td><td> Apr 3, 2025</td></tr>

Author: ```suggestion
<tr><td> Updated </td><td> Apr 15, 2025</td></tr>
```

More specifically, violations of a MUST or MUST NOT rule MUST result in an error. Violations of the
rules specified by any of the other all-capital terms MAY result in a warning, at discretion of the
implementation.

### Identifying package artifacts

The conda ecosystem distinguishes between two types of packages:

- Distributable package names: represented by a concrete, downloadable, extractable conda artifact.
- Virtual package names: not backed by any concrete artifact. They only exist on the client side.

#### Package names

A distributable package name MUST only consist of lowercase ASCII letters, numbers, hyphens,
periods and underscores. It MUST start with a letter, a number, or a single underscore. It MUST NOT
include two consecutive separators (hyphen, period, underscore).

Virtual package names MUST only consist of lowercase ASCII letters, numbers, hyphens, periods and
underscores. They MUST NOT use two consecutive separators, with one exception: they MUST start with
two underscores.

Distributable package names MUST match the following case-insensitive regex:
`^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`.

> Virtual package names MUST follow this regex: `^__[a-z0-9][._-]?([a-z0-9]+(\.|-|_|$))*$`.

Author: We want to disallow consecutive separators in the string, like `numpy--base` or `matplotlib.-extra`.

> Virtual package names MUST follow this regex: `^__[a-z0-9][._-]?([a-z0-9]+(\.|-|_|$))*$`.

Author: Maybe like this?
```re
^([a-z0-9]+|(_[a-z0-9]+))[._-]?([a-z0-9]+(\.|-|_|$))*$
```

> Virtual package names MUST follow this regex: `^__[a-z0-9][._-]?([a-z0-9]+(\.|-|_|$))*$`.

Reviewer: I've found these things are tricky and the negative lookahead was all that I could get to work when I was working on similar things.

`file:///home/username/channel`.

When present, each path component MUST only contain lowercase ASCII letters, numbers, underscores,
periods, and dashes. They MUST NOT start with a period or a dash. They SHOULD start and end with a
letter or a number. If present, each path component MUST match this regex:

```re
^[a-z0-9_][a-z0-9_.-]*$
```

For `file://`-based channel URLs, the path component rules MAY be understood as recommendations
only.

The maximum length of an individual path component in a channel base URL MUST NOT exceed 128
characters. The maximum length of a channel base URL SHOULD NOT exceed 256 characters.

To avoid ambiguous `MatchSpec` grammar, the last path component of a channel base URL SHOULD NOT
match any `subdir` identifiers. If it does, the behavior in this ambiguous case is not defined
and implementation dependent.

#### Channel names

For convenience, the channel _name_ is defined as the concatenation of `scheme`, `authority` and
`path` components of a channel URL. At least one of `authority` or `path` SHOULD be present. In
their absence, the channel name MUST be considered empty, regardless the scheme. Empty channel

> names SHOULD NOT be used.

Reviewer 4: This deviates, e.g., from how `conda`'s implementation uses (or at least used to use in the past) the term `name`.
E.g., for `conda.models.channel.Channel` a `https://conda.anaconda.org/conda-forge/label/main` would (using the default channel alias) yield `conda-forge/label/main` as the `Channel.name`.
One reason for this that `conda` itself has (had?) no concept of "label" at all but only distribution related tooling (e.g., `conda.anaconda.org`'s server or tools like `anaconda-client` or `conda-smithy`) use that concept.

This is not to say it does not make sense to include the "label" concept at all -- it's certainly good to have this distribution-specific part specified since it's widely used, e.g., in `conda-forge`.
We should just be careful to align meanings of channel names, labels, etc. in specs and tooling (which this proposal is/can be used to get it going, of course).

  authority, with the provided channel name taken as the rest of the path component. At the time of
  this CEP's writing, most tools assume the default URL scheme and authority to be
  `https://conda.anaconda.org`.

#### Subdir names

Channel subdir names MUST either be the literal `noarch` or a string following the syntax
`{os}-{arch}`, where `{os}` and `{arch}` MUST only consist of lowercase ASCII letters and numbers.
Non-`noarch` subdirs MUST match this regex: `^[a-z0-9]+-[a-z0-9]+$`.

The maximum length of a subdir name MUST NOT exceed 32 characters.

#### Label names

Channel label names MUST only consist of ASCII letters, digits, underscores, hyphens, forward
slashes, periods, and whitespace. They MUST start with a letter. They MUST match this regex:
`^[a-zA-Z][0-9a-zA-Z_\-\./]*$`. The last `/`-delimited component of a label
SHOULD NOT match any `subdir` identifier. If it does, the behavior in this ambiguous
case is undefined and implementation dependent.

The label `nolabel` is reserved and MUST only be used for conda packages which have no other
labels. In other words, in the space of labels, the empty set is represented by the labels

> `nolabel`.

Reviewer 4: I'm not aware of `nolabel` reservation -- is this an existing thing for `conda.anaconda.org` or is it something to do with the OCI work mentioned in the comments?

Author: Related to OCI.
